### ISSUE TITLE
Setup failed for custom integration 'dbuezas_eq3btsmart' with 2024.2.0b0

### ISSUE THREAD

**Issue by AnHardt** [dbuezas/eq3btsmart#106]
'''
Core
**2024.2.0b0** && **2024.2.0b1** && **2024.2.0b2** && **2024.2.0b3** && **2024.2.0b4**
Supervisor
2024.01.1
Operating System
11.5.rc3
'''
v6.3.0
'''
Logger: homeassistant.setup
Source: setup.py:251
First occurred: 12:58:53 (1 occurrences)
Last logged: 12:58:53

Setup failed for custom integration 'dbuezas_eq3btsmart': Unable to import component: No module named 'imp'
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/setup.py", line 251, in _async_setup_component
    component = integration.get_component()
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/loader.py", line 827, in get_component
    ComponentProtocol, importlib.import_module(self.pkg_path)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/importlib/__init__.py", line 90, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 994, in exec_module
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "/config/custom_components/dbuezas_eq3btsmart/__init__.py", line 5, in <module>
    from .python_eq3bt.eq3bt.eq3btsmart import Thermostat
  File "/config/custom_components/dbuezas_eq3btsmart/python_eq3bt/eq3bt/__init__.py", line 2, in <module>
    from .structures import *
  File "/config/custom_components/dbuezas_eq3btsmart/python_eq3bt/eq3bt/structures.py", line 4, in <module>
    from construct import (
  File "/usr/local/lib/python3.12/site-packages/construct/__init__.py", line 22, in <module>
    from construct.core import *
  File "/usr/local/lib/python3.12/site-packages/construct/core.py", line 3, in <module>
    import struct, io, binascii, itertools, collections, pickle, sys, os, tempfile, hashlib, importlib, imp
ModuleNotFoundError: No module named 'imp'

'''

**Comment by fightforlife** [comment]
Same issue here, following this.

**Comment by EuleMitKeule** [comment]
This is caused by the `imp` module being dropped in Python 3.12. I did a quick research and found that `construct` actually shouldn't be using `imp` anymore (see https://github.com/construct/construct/issues/862).
Could you find out which version of construct is installed in your environment?

**Comment by EuleMitKeule** [comment]
You could try to manually update `construct` to the latest version where `imp` is not used. Alternatively you can try out [my fork](https://github.com/eulemitkeule/eq3btsmart) of this integration where I updated the dependencies and this error should not occur.

**Comment by dbuezas** [comment]
I haven't updated Home Assistant yet (busy), I'll make an update as soon as I free some time

**Comment by killwarez** [comment]
Hi, have I faced the same issue? I lost connection to all my thermostats after update to 2024.1.6/11.4. They are discoverable when removed from integrations, but unavailable when added.

Core
2024.1.6
Operating System
11.4

`2024-02-04 07:07:52.181 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [eq3 Microwave] Querying the device..
2024-02-04 07:08:07.507 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.config_flow] Discovered eQ3 thermostat using bluetooth: <habluetooth.models.BluetoothServiceInfoBleak object at 0x7f49650b55c0>, CC-RT-BLE
2024-02-04 07:08:07.511 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.config_flow] Discovered eQ3 thermostat using bluetooth: <habluetooth.models.BluetoothServiceInfoBleak object at 0x7f4964e37bc0>, CC-RT-BLE
2024-02-04 07:08:08.836 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.config_flow] Discovered eQ3 thermostat using bluetooth: <habluetooth.models.BluetoothServiceInfoBleak object at 0x7f495cd42ac0>, CC-RT-BLE
2024-02-04 07:08:08.863 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.config_flow] Discovered eQ3 thermostat using bluetooth: <habluetooth.models.BluetoothServiceInfoBleak object at 0x7f494cf3a8c0>, CC-RT-BLE
2024-02-04 07:08:25.111 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [eq3 Microwave] Broken connection [retry 1/14]: eq3 Microwave - 00:1A:22:08:03:58: Failed to connect after 2 attempt(s): failed to discover services, device disconnected
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/bleak_retry_connector/__init__.py", line 363, in establish_connection
    await client.connect(
  File "/usr/local/lib/python3.11/site-packages/habluetooth/wrappers.py", line 295, in connect
    connected = await super().connect(**kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/bleak/__init__.py", line 605, in connect
    return await self._backend.connect(**kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/bleak/backends/bluezdbus/client.py", line 268, in connect
    await self.get_services(
  File "/usr/local/lib/python3.11/site-packages/bleak/backends/bluezdbus/client.py", line 656, in get_services
    self.services = await manager.get_services(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/bleak/backends/bluezdbus/manager.py", line 644, in get_services
    await self._wait_for_services_discovery(device_path)
  File "/usr/local/lib/python3.11/site-packages/bleak/backends/bluezdbus/manager.py", line 779, in _wait_for_services_discovery
    raise BleakError("failed to discover services, device disconnected")
bleak.exc.BleakError: failed to discover services, device disconnected

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/config/custom_components/dbuezas_eq3btsmart/python_eq3bt/eq3bt/bleakconnection.py", line 176, in _async_make_request_try
    conn = await self.async_get_connection()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/dbuezas_eq3btsmart/python_eq3bt/eq3bt/bleakconnection.py", line 94, in async_get_connection
    self._conn = await establish_connection(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/bleak_retry_connector/__init__.py", line 472, in establish_connection
    _raise_if_needed(name, device.address, exc)
  File "/usr/local/lib/python3.11/site-packages/bleak_retry_connector/__init__.py", line 339, in _raise_if_needed
    raise BleakConnectionError(msg) from exc
bleak_retry_connector.BleakConnectionError: eq3 Microwave - 00:1A:22:08:03:58: Failed to connect after 2 attempt(s): failed to discover services, device disconnected
2024-02-04 07:08:31.896 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.config_flow] Discovered eQ3 thermostat using bluetooth: <habluetooth.models.BluetoothServiceInfoBleak object at 0x7f4947be6740>, CC-RT-BLE
2024-02-04 07:08:31.897 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [eq3 Microwave] Connected
`

**Comment by fightforlife** [comment]
> This is caused by the `imp` module being dropped in Python 3.12. I did a quick research and found that `construct` actually shouldn't be using `imp` anymore (see [construct/construct#862](https://github.com/construct/construct/issues/862)). Could you find out which version of construct is installed in your environment?

I am using the official docker image "beta" and it is using 
`Requirement already satisfied: construct in /usr/local/lib/python3.12/site-packages (2.10.68)`

Updating manually to 2.10.70 (latest) in the docker gives a error message since construct-typing is requiring construct 2.10.68.
I updated eitherway to 2.10.70 and this integration is working again. 

@EuleMitKeule: I tried your fork/rewrite of this integration with btproxy, but had major problems connecting to my eq3. Will create an issue in your repo, when I get a glimpse of the reason. (the first conenction after setup works, but then all further connections fail)

**Comment by EuleMitKeule** [comment]
> since construct-typing is requiring construct 2.10.68

The latest release in this repo does not use `construct-typing`, so this stems from some other component in Home Assistant probably.
If updating to the latest version of `construct` fixed the problem, that would mean that some other integration you have installed is probably depending on a very outdated version of `construct`. If you find out which integration is at fault, you could create an issue in that integration's repo.

> I tried your fork/rewrite of this integration with btproxy, but had major problems connecting to my eq3. Will create an issue in your repo, when I get a glimpse of the reason. (the first conenction after setup works, but then all further connections fail)

Weird, this should definetely not happen at least in the latest version. Would be nice if you could create an issue and submit logs.

**Comment by AnHardt** [comment]
For my system updating 'custom_components/dbuezas_eq3btsmart/manifest.json'

from
``` 
  "requirements": ["construct==2.10.56"],
```
to 
``` 
  "requirements": ["construct==2.10.70"],
``` 
worked without any further errors.
Its working now with  Core 2024.2.0b5

~~As a side effect also 'wonderground pws' is running again.~~ (coincident, not corelated)

**Comment by dbuezas** [comment]
Thanks @AnHardt, I applied your correction and released v6.3.1.

Please close this issue if the latest release fixes it

**Comment by AnHardt** [comment]
For me fixed by PR#107 in release V6.3.1.
